huaweicloud/services/ims/resource_huaweicloud_images_image_share.go

```diff
@@ -0,0 +1,280 @@
+	}
+	d.SetId(resourceId)
+
+	return resourceImsImageShareRead(ctx, d, meta)
+}
+
+func resourceImsImageShareUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	cfg := meta.(*config.Config)
+	region := cfg.GetRegion(d)
+
+	if d.HasChange("target_project_ids") {
+		oProjectIdsRaw, nProjectIdsRaw := d.GetChange("target_project_ids")
+		shareProjectIds := nProjectIdsRaw.(*schema.Set).Difference(oProjectIdsRaw.(*schema.Set))
+		unShareProjectIds := oProjectIdsRaw.(*schema.Set).Difference(nProjectIdsRaw.(*schema.Set))
+		if shareProjectIds.Len() > 0 {
+			err := shareImages(ctx, d, cfg, region, shareProjectIds.List())
+			if err != nil {
+				return diag.FromErr(err)
+			}
+		}
+		if unShareProjectIds.Len() > 0 {
+			err := unShareImages(ctx, d, cfg, region, unShareProjectIds.List())
+			if err != nil {
+				return diag.FromErr(err)
+			}
+		}
+	}
+
+	return resourceImsImageShareRead(ctx, d, meta)
+}
+
+func resourceImsImageShareRead(_ context.Context, _ *schema.ResourceData, _ interface{}) diag.Diagnostics {
+	return nil
+}
+
+func resourceImsImageShareDelete(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	cfg := meta.(*config.Config)
+	region := cfg.GetRegion(d)
+
+	projectIds := d.Get("target_project_ids")
+	err := unShareImages(ctx, d, cfg, region, projectIds.(*schema.Set).List())
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	d.SetId("")
+
+	return nil
+}
+
+func shareImages(ctx context.Context, d *schema.ResourceData, cfg *config.Config, region string, projectIds []interface{}) error {
+	// createImageShare: create IMS image share
+	var (
+		createImageShareHttpUrl = "v1/cloudimages/members"
+		createImageShareProduct = "ims"
+	)
+	createImageShareClient, err := cfg.NewServiceClient(createImageShareProduct, region)
+	if err != nil {
+		return fmt.Errorf("error creating IMS Client: %s", err)
+	}
+
+	createImageSharePath := createImageShareClient.Endpoint + createImageShareHttpUrl
+
+	createImageShareOpt := golangsdk.RequestOpts{
+		KeepResponseBody: true,
+		OkCodes: []int{
+			200,
+		},
+	}
+	createImageShareOpt.JSONBody = utils.RemoveNil(buildCreateImageShareBodyParams(d, projectIds))
+	createImageShareResp, err := createImageShareClient.Request("POST", createImageSharePath, &createImageShareOpt)
+	if err != nil {
+		return fmt.Errorf("error creating IMS image share: %s", err)
+	}
+
+	createImageShareRespBody, err := utils.FlattenResponse(createImageShareResp)
+	if err != nil {
+		return err
+	}
+
+	jobId, err := jmespath.Search("job_id", createImageShareRespBody)
+	if err != nil {
+		return fmt.Errorf("error creating IMS image share: job_id is not found in API response")
+	}
+
+	err = waitForJobSuccess(ctx, d, createImageShareClient, jobId.(string), schema.TimeoutCreate)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func buildCreateImageShareBodyParams(d *schema.ResourceData, projectIds []interface{}) map[string]interface{} {
```

Comment:
please try to merge buildCreateImageShareBodyParams and buildDeleteImageShareBodyParams

Reply:
done